Add docs for CHT Monitoring setup (#1020)
medic/cht-docs: content/en/apps/guides/hosting/monitoring.md

The solution provides both an overview dashboard as well as a detail dashboard.  Here is a portion of the overview dashboard:

![Screenshot of Grafana Dashboard showing data from Prometheus](monitoring.and.alerting.screenshot.png)

[Prometheus supports](https://prometheus.io/docs/concepts/metric_types/) four metric types: Counter, Gauge, Histogram, and Summary.  Currently, the CHT only provides Counter and Gauge type metrics. When building panels for Grafana dashboards, [Prometheus Functions](https://prometheus.io/docs/prometheus/latest/querying/functions/) can be used to manipulate the metric data. Refer to the [Grafana Documentation](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/best-practices/) for best practices on building dashboards.

### Prerequisites

- [Docker](https://docs.docker.com/install/)
- [Docker Compose](https://docs.docker.com/compose/install/)
- [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- URL(s) of the CHT instance(s) 

### Setup

> Pre-provisioned dashboards are included in the `CHT` folder in the Grafana UI should not be updated directly. 

Author: ah - agreed! fixed in 3b8865e

 I tried to find a home for all the original content, but we can remove for now.

> Pre-provisioned dashboards are included in the `CHT` folder in the Grafana UI should not be updated directly. 

Reviewer: :laughing: So, I did not actually mean to post that comment as-is!  I created it as I went through the review and meant to come back after reading the rest of it to see if there was a better place to put this. But, I forgot to come back to this. Oh well....

These instructions have been tested against Ubuntu, but should work against any OS that meets the prerequisites. They follow a happy path assuming you need to only set a secure password and specify the URL(s) to monitor:

1. Run the following commands to clone this repository, initialize your `.env` file, create a secure password and create your data directories:

    ```sh
    cd ~
    git clone https://github.com/medic/cht-monitoring.git
    cd cht-monitoring
    cp cht-instances.example.yml cht-instances.yml
    cp grafana/grafana.example.ini grafana/grafana.ini
    mkdir -p grafana/data && mkdir  -p prometheus/data 

>     apt install -y wamerican

Author: ah - good call!  I just learned you can install this file the other month when I did the single/multi node cht stuffs.  Hope my "always use a secure passphrase" preferences aren't too dogmatic!

fixed in c6334d2

    cp .env.example .env
    password=$(shuf -n7 /usr/share/dict/words --random-source=/dev/random | tr '\n' '-' | tr -d "'" | cut -d'-' -f1,2,3,4,5,6,7)
    sed -i -e "s/password/$password/g" .env

>     echo;echo "Setup complete! To log into Grafana in the browser:";echo 

Author: hah!  agreed. fixed in a513d60

    echo "    username: medic"
    echo "    password: ${password}";echo
    ```
   Note that in step 4 below you'll need the username and password which is printed after you run the above command.

> 2. Edit the `cht-instances.yml` file to have the URLs of your CHT instances. You may include as many URLs of CHT instances as you like. To preserve data consistency parameters in this file should NOT be adjusted after it is initially set. Be sure to not include a trailing slash (`/`).

Author: fixed in 5776efd

    ```

4. Grafana is available at [http://localhost:3000](http://localhost:3000). See the output from step 1 for your username and password.

If you would like to do more customizing of your deployment, see ["Addition Configuration"](#additional-configuration).

### Upgrading

Before upgrading, you should back up both your current configuration settings and your Prometheus time-series data.

#### Prometheus, Grafana and JSON Exporter

To upgrade these dependencies, update the version numbers set in your `.env` file (or leave them set to `latest`).  Then run the following commands:

```shell
docker compose pull
docker compose up -d
```

#### CHT Monitoring Config

When you see a new version in the [GitHub repository](https://github.com/medic/cht-monitoring), first review the release notes and upgrade instructions. Then, run the following commands to deploy the new configuration:

1. Note the version number to derive the branch to pull.  For example,  `1.1.0` would be `1.1.x`
2. Run this command to update your instance and restart it. Be sure to replace `BRANCH_NAME` with the value from the first step:
   ```shell
   cd ~/cht-monitoring
   git fetch
   git checkout BRANCH_NAME

>    ```shell

Reviewer: Since new versions of the config could include additions to the docker-compose file, I think this should include:

```shell
docker compose pull
docker compose up -d --remove-orphans
```

Author: ah - ok - I had removed that thinking we might not want to upgrade grafana/prometheus when we're just trying to update the cht-monitoring part.  I guess that should come in a separate commit to pin to a specific release vs `latest` like we have no.  That works!

I added a restart in 659ace5 and then I added `pull` just now in 1b34f0a.  Lemme know if we require the `compose up -d --remove-orphans` , assuming not though!

Reviewer: So, I might be wrong, but I don't think `restart` will actually update the image used (even if you have done a `pull` and gotten a new version) if you have existing containers (it just keeps the same containers).

The case I am trying to cover is if we ever update our base docker-compose config to add or remove a service (e.g. removing json-exporter since we don't need it any more).  In that case we also want the `--remove-orphans` since the json-exporter service will not be in the new version of the docker-compose file.

Author: gotcha!  Added `--remove-orphans` to both calls and added a `docker compose down` before the `up` to not have the `restart` any more in e37e1be4.  Thanks for seeing this through!

### Additional Configuration

#### Prometheus Retention and Storage

By default, historical monitoring data will be stored in Prometheus (`PROMETHEUS_DATA` directory) for 60 days (configurable by `PROMETHEUS_RETENTION_TIME`). A longer retention time can be configured to allow for longer-term analysis of the data.  However, this will increase the size of the Prometheus data volume.  See the [Prometheus documentation](https://prometheus.io/docs/prometheus/latest/storage/) for more information.

Local storage is not suitable for storing large amounts of monitoring data. If you intend to store multiple years worth of metrics, you should consider integrating Prometheus with a [Remote Storage](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage/).

#### Alerts

> This configuration includes number of pre-provisioned alerts.  Specific alerting rules (and other contact points) can be set in the Grafana UI.  

Reviewer: ```suggestion
This configuration includes number of pre-provisioned alerts.  Additional alerting rules (and other contact points) can be set in the Grafana UI.  
```

> See the both Grafana [high level alert Documentation](https://grafana.com/docs/grafana/latest/alerting/) and  [provisioning alerts in the UI](https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/file-provisioning/#provision-alert-rules) for more information on how to edit or remove these provisioned alerts.

Reviewer: ```suggestion
See both the Grafana [high level alert Documentation](https://grafana.com/docs/grafana/latest/alerting/) and [provisioning alerts in the UI](https://grafana.com/docs/grafana/latest/alerting/set-up/provision-alerting-resources/file-provisioning/#provision-alert-rules) for more information on how to edit or remove these provisioned alerts.
```

Additionally, you can configure where these alerts are sent.  Two likely options are Email and Slack.

##### Email

> To support sending email alerts from Grafana, you must update the `smtp` section of your `grafana/grafana.ini` file with your SMTP server configuration.  Ten, in the web interface, add the desired recipient email addresses in the `grafana-default-email` contact point settings.

Reviewer: ```suggestion
To support sending email alerts from Grafana, you must update the `smtp` section of your `grafana/grafana.ini` file with your SMTP server configuration.  Then, in the web interface, add the desired recipient email addresses in the `grafana-default-email` contact point settings.
```

##### Slack

> Slack alerts can be configured within the Grafana web GUI for the specific panel you would like to alert on.  Find the panel you want to alert on and edit it.  In the bottom half of the screen, choose the "Alert" tab. Then click "Create alert fule from this panel".

Reviewer: ```suggestion
Slack alerts can be configured within the Grafana web GUI for the specific rules you would like to alert on.
```

>  Find the panel you want to alert on and edit it.  In the bottom half of the screen, choose the "Alert" tab. Then click "Create alert fule from this panel".

This feels like odd information to include in the Slack section since it is simply general info on how to create an alert from a panel. They should be able to trigger Slack notifications for existing alerts (and email notifs for alerts created from panels...).

Author: Agreed!
